### PR TITLE
New modmail

### LINKED ISSUE #56: Add support for new modmail API endpoints
https://www.reddit.com/r/redditdev/comments/5jf4yg/api_new_modmail/

### PR DESCRIPTION
**Pull request by SpyTec**
~~This is currently in-progress and nowhere near finished. I wanted to get more centralized feedback on the direction and scope of the implementation instead of cluttering up #56 with snippets.~~

Once this is finished, it fixes #56. 

**Edit:** Ready for review

### CODE REVIEW

File: src/snoowrap.js [hunk 1]
```diff
@@ -695,270 +728,374 @@ const snoowrap = class snoowrap {
...
+        return this._newObject('Listing', {
+          after: null,
+          before: null,
+          children: Object.values(response.subreddits).map(s => {
+            return this._newObject('Subreddit', s);
+          })
+        });
+      }
+    });
+  }
```
Comment (SpyTec) [inline src/snoowrap.js]: @not-an-aardvark Is there a better way to do this? The format received from the endpoint is as follows:

```json
{
  "subreddits": {
    "t5_xogi3": {
      "display_name": "SpyTecSnoowrapTesting",
      "name": "SpyTecSnoowrapTesting",
      "lastUpdated": "2019-03-19T09:45:13.994306+00:00",
      "keyColor": "",
      "subscribers": 1,
      "id": "t5_xogi3",
      "icon": ""
    },
    "t5_xog8s": {
      "display_name": "u/SpyTecSnoowrap",
      "name": "u_SpyTecSnoowrap",
      "lastUpdated": null,
      "keyColor": "",
      "subscribers": 0,
      "id": "t5_xog8s",
      "icon": "https://www.redditstatic.com/avatars/avatar_default_20_46D160.png"
    }
  }
}
```

**Comment by SpyTec** [comment]
Ready for review

**Comment by not-an-aardvark** [comment]
Sorry about the delay, I've been a bit busy recently. Just letting you know that I've seen this and will review when I get the chance.

**Comment by SpyTec** [comment]
@not-an-aardvark have any time in the upcoming weeks? Would be nice with review while my mind still remembers the code :)

File: src/snoowrap.js [hunk 2]
```diff
@@ -695,270 +728,478 @@ const snoowrap = class snoowrap {
...
+   * @example
+   *
+   * r.createModeratorDiscussion({
+   *   body: 'test body',
+   *   subject: 'test subject',
+   *   srName: 'AskReddit'
+   * }).then(console.log)
+   * // ModmailConversation { messages: [...], objIds: [...], subject: 'test subject', ... }
+   */
+  createModeratorDiscussion ({
```
Comment (not-an-aardvark) [inline src/snoowrap.js]: Minor: Maybe this should be called `createModmailDiscussion` for consistency with the other functions that use the term `modmail`.

File: src/snoowrap.js (hunk 1 above)
Comment (not-an-aardvark) [reply src/snoowrap.js]: If it's not paginated, would it make sense to just use an array instead of a Listing?

File: src/snoowrap.js (hunk 2 above)
Comment (SpyTec) [reply src/snoowrap.js]: Sounds good to me

File: src/snoowrap.js (hunk 1 above)
Comment (SpyTec) [reply src/snoowrap.js]: That makes sense, I’ll change it when I get the chance

Comment (SpyTec) [reply src/snoowrap.js]: Changed to be an array, tests have been updated as well

File: src/snoowrap.js (hunk 2 above)
Comment (SpyTec) [reply src/snoowrap.js]: Changed

**Comment by not-an-aardvark** [comment]
Hi, just checking in on this PR -- is it still a work in progress?

**Comment by SpyTec** [comment]
>Hi, just checking in on this PR -- is it still a work in progress?

@not-an-aardvark it’s ready for review. Did you notice anything odd?

**Comment by not-an-aardvark** [comment]
No, I had been looking at the comment at the top of the PR that says "This is currently in-progress and nowhere near finished." Sorry about the miscommunication -- I'll try to take another look in the next few days.

**Comment by SpyTec** [comment]
> No, I had been looking at the comment at the top of the PR that says "This is currently in-progress and nowhere near finished." Sorry about the miscommunication -- I'll try to take another look in the next few days.

No worries, I should've changed that too :)

**Comment by not-an-aardvark** [comment]
Sorry about the long delay, this is still on my radar. I'll review it over this weekend.

**Comment by SpyTec** [comment]
Added it to the subreddit. Not sure if `mail` is all the permissions it needs but if it doesn't work I can change it

Ideally we should have a script that can setup the testing environment for us. If I get the time I would like to try and work on this next when I get the time. This would need to rely on Snoowrap itself to set up the environment though.

**Comment by not-an-aardvark** [comment]
Thanks. All the tests are passing for me with the exception of `can mute a user`:

```
  1) snoowrap new modmail can mute a user:
     StatusCodeError: 400 - {"fields":[null],"explanation":null,"message":"Bad Request","reason":"CANT_RESTRICT_MODERATOR"}
      at new StatusCodeError (node_modules/request-promise/lib/errors.js:32:15)
      at node_modules/request-promise/lib/rp.js:70:37
      at tryCatcher (node_modules/bluebird/js/release/util.js:16:23)
      at Promise._settlePromiseFromHandler (node_modules/bluebird/js/release/promise.js:512:31)
      at Promise._settlePromise (node_modules/bluebird/js/release/promise.js:569:18)
      at Promise._settlePromiseCtx (node_modules/bluebird/js/release/promise.js:606:10)
      at Async._drainQueue (node_modules/bluebird/js/release/async.js:138:12)
      at Async._drainQueues (node_modules/bluebird/js/release/async.js:143:10)
      at Immediate.Async.drainQueues [as _onImmediate] (node_modules/bluebird/js/release/async.js:17:14)
      at processImmediate (internal/timers.js:439:21)
```

Do you know what might be happening there? If not, I can try to look into it (it's possible that it's a local condition).

**Comment by SpyTec** [comment]
My bad, missed out one permission for the bot. Should work now!

**Comment by not-an-aardvark** [comment]
Great, thanks! I'll merge this and try to get a release out.

**Comment by not-an-aardvark** [comment]
Published as v1.16.0.

**Comment by SpyTec** [comment]
@not-an-aardvark Seems like tests failed due to permissions, does it run on a different bot?

Also, do you have an IM platform you use where we talk?

**Comment by not-an-aardvark** [comment]
The tests on Travis should be using the `/u/snoowrap_testing` account, but I'm guessing I forgot to add the `mail` scope (since the scope didn't exist when the token was generated). I'll add it now.

We could use [Gitter](https://gitter.im/not-an-aardvark/snoowrap) (I set up a chatroom for this repository awhile ago, but it doesn't appear to have gotten any use). Gitter also supports private messages via the "person" icon on the left sidebar.

**Comment by markspolakovs** [comment]
@not-an-aardvark have the docs been updated? https://not-an-aardvark.github.io/snoowrap/Subreddit.html doesn't have `getNewModmailConversations`

**Comment by SpyTec** [comment]
> 
> 
> @not-an-aardvark have the docs been updated? https://not-an-aardvark.github.io/snoowrap/Subreddit.html doesn't have `getNewModmailConversations`

Seems like I forgot to add that. I've opened #199 for it

In the meantime you can call `snoowrap.getNewModmailConversations({entity: 'subredditName'})` which is more or less what the `Subreddit.getNewModmailConversations()` will do

**Comment by markspolakovs** [comment]
Oh right @SpyTec, thanks! Just wondering, are there any docs for the fields on `ModmailConversation` and `ModmailConversationAuthor`? I'm working on the TypeScript types for them now and a reference would be handy.

**Comment by SpyTec** [comment]
@markspolakovs We would at some point want to add either TS definition files or convert components to TypeScript as the project goes on if that's something you want to help out with 😉 

Unfortunately, it will be difficult to map what is returned from new modmail. From a rant I posted to a few friends:

>For instance, participant may not be there. Sometimes it is, sometimes it isn't. I don't know when it's there or when it isn't. It also moves around depending on if it's in list or detail view. Same with author

I can give JSON response for each endpoint though:

## /conversations
```
{
    "conversations":
    {
        "75hxt":
        {
            "isAuto": false,
            "objIds":
            [
                {
                    "id": "eg3v4",
                    "key": "messages"
                }
            ],
            "isRepliable": true,
            "lastUserUpdate": "2019-03-07T14:11:06.909685+00:00",
            "isInternal": false,
            "lastModUpdate": "2019-07-24T15:16:33.809212+00:00",
            "lastUpdated": "2019-07-24T15:16:33.809212+00:00",
            "authors":
            [
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "snoowrap_testing",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 50555952,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "SpyTec13",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 13118708,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "SpyTec13",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 13118708,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "SpyTec13",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 13118708,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "SpyTecSnoowrap",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 262998090119,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "SpyTecSnoowrap",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 262998090119,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "SpyTecSnoowrap",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 262998090119,
                    "isDeleted": false
                },
                {
                    "isMod": true,
                    "isAdmin": false,
                    "name": "SpyTecSnoowrap",
                    "isOp": false,
                    "isParticipant": false,
                    "isHidden": false,
                    "id": 262998090119,
                    "isDeleted": false
                },
                {
                    "isMod": false,
                    "isAdmin": false,
                    "name": "EliteDangerousBot",
                    "isOp": true,
                    "isParticipant": true,
                    "isHidden": false,
                    "id": 47524494,
                    "isDeleted": false
                }
            ],
            "owner":
            {
                "displayName": "SpyTecSnoowrapTesting",
                "type": "subreddit",
                "id": "t5_xogi3"
            },
            "id": "75hxt",
            "isHighlighted": false,
            "subject": "snoowrap test",
            "participant":
            {
                "isMod": false,
                "isAdmin": false,
                "name": "EliteDangerousBot",
                "isOp": true,
                "isParticipant": true,
                "isHidden": false,
                "id": 47524494,
                "isDeleted": false
            },
            "state": 1,
            "lastUnread": null,
            "numMessages": 27
        }
    },
    "messages":
    {
        "eg3v4":
        {
            "body": "<!-- SC_OFF --><div class=\"md\"><p>You have been <a href=\"https://www.reddithelp.com/en/categories/reddit-101/moderators/modmail-muting\">temporarily muted</a> from <a href=\"/r/SpyTecSnoowrapTesting\">r/SpyTecSnoowrapTesting</a>. You will not be able to message the moderators of <a href=\"/r/SpyTecSnoowrapTesting\">r/SpyTecSnoowrapTesting</a> for 72 hours.</p>\n</div><!-- SC_ON -->",
            "author":
            {
                "isMod": true,
                "isAdmin": false,
                "name": "snoowrap_testing",
                "isOp": false,
                "isParticipant": false,
                "isHidden": false,
                "id": 50555952,
                "isDeleted": false
            },
            "isInternal": false,
            "date": "2019-07-24T15:16:33.809212+00:00",
            "bodyMarkdown": "You have been [temporarily muted](https://www.reddithelp.com/en/categories/reddit-101/moderators/modmail-muting) from r/SpyTecSnoowrapTesting. You will not be able to message the moderators of r/SpyTecSnoowrapTesting for 72 hours.",
            "id": "eg3v4"
        }
    },
    "viewerId": "t2_7t6gk",
    "conversationIds":
    [
        "75hxt"
    ]
}
```

## /conversation/:id

```
{
    "conversation":
    {
        "isAuto": true,
        "objIds":
        [
            {
                "id": "ddfzq",
                "key": "messages"
            },
            {
                "id": "5p28r",
                "key": "modActions"
            }
        ],
        "isRepliable": true,
        "lastUserUpdate": null,
        "isInternal": false,
        "lastModUpdate": "2019-06-04T11:04:15.186318+00:00",
        "lastUpdated": "2019-06-04T11:04:15.186318+00:00",
        "authors":
        [
            {
                "isMod": true,
                "isAdmin": false,
                "name": "SpyTec13",
                "isOp": true,
                "isParticipant": false,
                "isHidden": true,
                "id": 13118708,
                "isDeleted": false
            },
            {
                "isMod": true,
                "isAdmin": false,
                "name": "SpyTec13",
                "isOp": true,
                "isParticipant": false,
                "isHidden": false,
                "id": 13118708,
                "isDeleted": false
            }
        ],
        "owner":
        {
            "displayName": "SpyTecSnoowrapTesting",
            "type": "subreddit",
            "id": "t5_xogi3"
        },
        "id": "8f5rc",
        "isHighlighted": false,
        "subject": "invitation to moderate /r/SpyTecSnoowrapTesting",
        "participant": {},
        "state": 2,
        "lastUnread": null,
        "numMessages": 1
    },
    "messages":
    {
        "ddfzq":
        {
            "body": "<!-- SC_OFF --><div class=\"md\"><p><strong>gadzooks! you are invited to become a moderator of <a href=\"/r/SpyTecSnoowrapTesting\">/r/SpyTecSnoowrapTesting: SpyTecSnoowrapTesting</a>!</strong></p>\n\n<p><em>to accept</em>, visit the <a href=\"/r/SpyTecSnoowrapTesting/about/moderators\">moderators page for /r/SpyTecSnoowrapTesting</a> and click &quot;accept&quot;.</p>\n\n<p><em>otherwise,</em> if you did not expect to receive this, you can simply ignore this invitation or report it.</p>\n</div><!-- SC_ON -->",
            "author":
            {
                "isMod": true,
                "isAdmin": false,
                "name": "SpyTec13",
                "isOp": true,
                "isParticipant": false,
                "isHidden": true,
                "id": 13118708,
                "isDeleted": false
            },
            "isInternal": false,
            "date": "2019-06-04T11:04:15.186318+00:00",
            "bodyMarkdown": "**gadzooks! you are invited to become a moderator of [/r/SpyTecSnoowrapTesting: SpyTecSnoowrapTesting](/r/SpyTecSnoowrapTesting)!**\n\n*to accept*, visit the [moderators page for /r/SpyTecSnoowrapTesting](/r/SpyTecSnoowrapTesting/about/moderators) and click \"accept\".\n\n*otherwise,* if you did not expect to receive this, you can simply ignore this invitation or report it.",
            "id": "ddfzq"
        }
    },
    "user":
    {
        "recentComments": {},
        "muteStatus":
        {
            "isMuted": false,
            "endDate": null,
            "reason": ""
        },
        "name": "snoowrap_testing",
        "created": "2016-01-25T06:26:36.059057+00:00",
        "banStatus":
        {
            "endDate": null,
            "reason": "",
            "isBanned": false,
            "isPermanent": false
        },
        "isSuspended": false,
        "isShadowBanned": false,
        "recentPosts": {},
        "recentConvos":
        {
            "8f5rc":
            {
                "date": "0001-01-01T00:00:00+00:00",
                "permalink": "https://mod.reddit.com/mail/perma/8f5rc",
                "id": "8f5rc",
                "subject": "invitation to moderate /r/SpyTecSnoowrapTesting"
            }
        },
        "id": "t2_u3l80"
    },
    "modActions":
    {
        "5p28r":
        {
            "date": "2019-07-24T18:35:00.822908+00:00",
            "actionTypeId": 2,
            "id": "5p28r",
            "author":
            {
                "name": "SpyTec13",
                "isMod": true,
                "isAdmin": false,
                "isHidden": false,
                "id": 13118708,
                "isDeleted": false
            }
        }
    }
}
```

## /conversations/count
```
{
    "highlighted": 0,
    "notifications": 0,
    "archived": 0,
    "new": 0,
    "inprogress": 0,
    "mod": 0
}
```

**Comment by markspolakovs** [comment]
@SpyTec thanks for that! Yeah, I remember implementing this API when it was first a thing - one line somewhere went `name: message.user.user.name`, prefixed by `// what the f**k?`.

I'll try to bang out a PR to DefinitelyTyped in the next few days.

**Pull request merged**